# Cleaning Enron Dataset
In this notebook, we will explore different aspects of the enron data set to infer different information. We will be analyzing and simplifying every column to retrieve different information.

Let's begin by importing some of the required modules.

In [1]:
import email, re
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
%matplotlib inline

Let's import the processed pickle from previous notebook and check it.

In [35]:
data = pd.read_pickle('data/emails-processed.pkl')
data.head()

Date  \
Message-ID                                                          
<18782981.1075855378110.JavaMail.evans@thyme> 2001-05-14 23:39:00   
<15464986.1075855378456.JavaMail.evans@thyme> 2001-05-04 20:51:00   
<24216240.1075855687451.JavaMail.evans@thyme> 2000-10-18 10:00:00   
<13505866.1075863688222.JavaMail.evans@thyme> 2000-10-23 13:13:00   
<30922949.1075863688243.JavaMail.evans@thyme> 2000-08-31 12:07:00   

                                                                    From  \
Message-ID                                                                 
<18782981.1075855378110.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<15464986.1075855378456.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<24216240.1075855687451.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<13505866.1075863688222.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<30922949.1075863688243.JavaMail.evans@thyme>  (phillip.allen@enron.com)   

                                                                      To  \
Message-ID                                                                 
<18782981.1075855378110.JavaMail.evans@thyme>     (tim.belden@enron.com)   
<15464986.1075855378456.JavaMail.evans@thyme>  (john.lavorato@enron.com)   
<24216240.1075855687451.JavaMail.evans@thyme>   (leah.arsdall@enron.com)   
<13505866.1075863688222.JavaMail.evans@thyme>    (randall.gay@enron.com)   
<30922949.1075863688243.JavaMail.evans@thyme>     (greg.piper@enron.com)   

                                                 Subject           X-From  \
Message-ID                                                                  
<18782981.1075855378110.JavaMail.evans@thyme>             Phillip K Allen   
<15464986.1075855378456.JavaMail.evans@thyme>        Re:  Phillip K Allen   
<24216240.1075855687451.JavaMail.evans@thyme>   Re: test  Phillip K Allen   
<13505866.1075863688222.JavaMail.evans@thyme>             Phillip K Allen   
<30922949.1075863688243.JavaMail.evans@thyme>  Re: Hello  Phillip K Allen   

                                                                                            X-To  \
Message-ID                                                                                         
<18782981.1075855378110.JavaMail.evans@thyme>           Tim Belden <Tim Belden/Enron@EnronXGate>   
<15464986.1075855378456.JavaMail.evans@thyme>  John J Lavorato <John J Lavorato/ENRON@enronXg...   
<24216240.1075855687451.JavaMail.evans@thyme>                                   Leah Van Arsdall   
<13505866.1075863688222.JavaMail.evans@thyme>                                      Randall L Gay   
<30922949.1075863688243.JavaMail.evans@thyme>                                         Greg Piper   

                                              X-cc X-bcc  \
Message-ID                                                 
<18782981.1075855378110.JavaMail.evans@thyme>              
<15464986.1075855378456.JavaMail.evans@thyme>              
<24216240.1075855687451.JavaMail.evans@thyme>              
<13505866.1075863688222.JavaMail.evans@thyme>              
<30922949.1075863688243.JavaMail.evans@thyme>              

                                                                                        X-Folder  \
Message-ID                                                                                         
<18782981.1075855378110.JavaMail.evans@thyme>  \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Se...   
<15464986.1075855378456.JavaMail.evans@thyme>  \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Se...   
<24216240.1075855687451.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   
<13505866.1075863688222.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   
<30922949.1075863688243.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   

                                              X-Origin  \
Message-ID                                               
<18782981.1075855378110.JavaMail

The describe() method will provide us with a simple summary of the data set. We can see the count of the non-empty values present in each columns.

In [3]:
data.describe()

Date                  From                      To  \
count                517401                517401                  495554   
unique               224122                 20328                   54748   
top     2001-06-27 23:02:00  (kay.mann@enron.com)  (pete.davis@enron.com)   
freq                   1118                 16735                    9155   
first   1980-01-01 00:00:00                   NaN                     NaN   
last    2044-01-04 22:48:58                   NaN                     NaN   

       Subject    X-From    X-To    X-cc   X-bcc  \
count   517401    517372  517372  517372  517372   
unique  159290     27980   73552   33701     132   
top             Kay Mann                           
freq     19187     16324    9124  388486  517197   
first      NaN       NaN     NaN     NaN     NaN   
last       NaN       NaN     NaN     NaN     NaN   

                                                X-Folder    X-Origin  \
count                                             517372      517372   
unique                                              5335         259   
top     \Kay_Mann_June2001_1\Notes Folders\All documents  Kaminski-V   
freq                                                6639       26995   
first                                                NaN         NaN   
last                                                 NaN         NaN   

         X-FileName                                            content  \
count        517372                                             517401   
unique          429                                             249025   
top     vkamins.nsf  As you know, Enron Net Works (ENW) and Enron G...   
freq          24029                                                112   
first           NaN                                                NaN   
last            NaN                                                NaN   

              user  
count       517401  
unique         150  
top     kaminski-v  
freq         28465  
first          NaN  
last           NaN

From above table, we can see that there are total `517401` number of records in total. Many fields seem to be affected with null values, particularly the `X-bcc`, `X-cc` and `X-To` columns. Many emails have null values in the `To` fields, which cannot be used in our analysis since we cannot deduct who the mail was sent to.

In [4]:
data[data['To'].isnull()][:1]

Date  \
Message-ID                                                          
<15201149.1075855691021.JavaMail.evans@thyme> 2000-05-01 10:56:00   

                                                                    From  \
Message-ID                                                                 
<15201149.1075855691021.JavaMail.evans@thyme>  (phillip.allen@enron.com)   

                                                 To            Subject  \
Message-ID                                                               
<15201149.1075855691021.JavaMail.evans@thyme>  None  Re: DSL- Installs   

                                                        X-From  \
Message-ID                                                       
<15201149.1075855691021.JavaMail.evans@thyme>  Phillip K Allen   

                                                                     X-To  \
Message-ID                                                                  
<15201149.1075855691021.JavaMail.evans@thyme>  Circuit Provisioning@ENRON   

                                              X-cc X-bcc  \
Message-ID                                                 
<15201149.1075855691021.JavaMail.evans@thyme>              

                                                                                      X-Folder  \
Message-ID                                                                                       
<15201149.1075855691021.JavaMail.evans@thyme>  \Phillip_Allen_Dec2000\Notes Folders\'sent mail   

                                              X-Origin  X-FileName  \
Message-ID                                                           
<15201149.1075855691021.JavaMail.evans@thyme>  Allen-P  pallen.nsf   

                                                                                         content  \
Message-ID                                                                                         
<15201149.1075855691021.JavaMail.evans@thyme>  No one will be home on 5/11/00 to meet DSL ins...   

                                                  user  
Message-ID                                              
<15201149.1075855691021.JavaMail.evans@thyme>  allen-p

The above example shows that the mail was sent by phillip.allen@enron.com, but the receiver is not mentioned. Such emails cannot be used in network analysis of the users since we cannot define a node using the null values.

Let's group the records by the Date the email was sent on. The following table shows the number of emails sent on each date.

In [5]:
emails_by_date = data.groupby('Date').count()
emails_by_date[:5]

From   To  Subject  X-From  X-To  X-cc  X-bcc  X-Folder  \
Date                                                                           
1980-01-01 00:00:00   522  316      522     522   522   522    522       522   
1986-04-26 15:22:07     1    1        1       1     1     1      1         1   
1986-05-01 14:37:34     1    1        1       1     1     1      1         1   
1997-01-01 17:22:03     1    1        1       1     1     1      1         1   
1997-03-03 11:00:00     3    0        3       3     3     3      3         3   

                     X-Origin  X-FileName  content  user  
Date                                                      
1980-01-01 00:00:00       522         522      522   522  
1986-04-26 15:22:07         1           1        1     1  
1986-05-01 14:37:34         1           1        1     1  
1997-01-01 17:22:03         1           1        1     1  
1997-03-03 11:00:00         3           3        3     3

It seems that `522` emails were sent on `1980-01-01`, but the Enron corporation was formed on `1985`. It seems that the Date column is also filled with mis-labeled data.

There's a lot to do before we can use the dataset. Firstly, we'll use Regular expressions to extract the user name, the domain and the top level domains from each email. Following is the domain and email user name regex which will be combined to obtain the regular expression for the email.

In [6]:
domainRegex = r"(?P<full_domain>(?P<domain>(?:(?!-)(?:xn--|_{1,1})?[a-z0-9-]{0,61}[a-z0-9]{1,1}\.)*)(?P<tld>(?:xn--)?[a-z0-9\-]{1,61}|[a-z0-9-]{1,30}\.[a-z]{2,}))"
emailUsernameRegex = r"(?P<username>[a-zA-Z0-9.!#$%&*+/=?^_`|}~-]+)"
emailRegex = emailUsernameRegex + "@" + domainRegex + r""

Then, we will need to compile the regular expressions to use it. We will use the python `re` module for it.

In [7]:
email_regex = re.compile(emailRegex)

Let's create a python dictionary to temporarily store values extracted from the emails using the regular expression. The keys of the dictionary will be the extracted item from the email.

In [56]:
emails_data = {
    "from_username": [],
    "from_full_domain": [],
    "from_domain": [],
    "from_tld": [],
    "to_username": [],
    "to_full_domain": [],
    "to_domain": [],
    "to_tld": []
}

First, lets apply the regular expression on the From field of the emails. Any null values or anything not an email will be replaced with empty strings.

In [59]:
for emails in data.From:
    for email in emails:
        try:
            items = list(email_regex.search(email).groups())
        except AttributeError:
            items = ['', '', '', '']
        emails_data['from_username'].append(items[0])
        emails_data['from_full_domain'].append(items[1])
        emails_data['from_domain'].append(items[2][:-1])
        emails_data['from_tld'].append(items[3])

Let's repeat the process for the To field. We'll need to some extra work to manage the multiple emails in the To fields.

In [57]:
for emails in data.To:
    try:
        if len(emails)>1:
            username_list, fdomain_list, domain_list, tld_list = [], [], [], []
            for email in emails:
                items = list(email_regex.search(str(email)).groups())
                username_list.append(items[0])
                fdomain_list.append(items[1])
                domain_list.append(items[2][:-1])
                tld_list.append(items[3])
            items = [username_list, fdomain_list, domain_list, tld_list]
        else:
            for email in emails:
                items = list(email_regex.search(str(email)).groups())
    except (TypeError, AttributeError):
        items = ['', '', '', '']
        
    emails_data['to_username'].append(items[0])
    emails_data['to_full_domain'].append(items[1])
    emails_data['to_domain'].append(items[2][:-1])
    emails_data['to_tld'].append(items[3])

After extracting the respective values, we can then create a pandas DataFrame using the dictionary using the `pd.DataFrame.from_dict()` method. The column names will be the key names from the dictionary. Then, we'll set the index same as that of our `data` dataframe to help in concatenation.

In [61]:
emails_df = pd.DataFrame.from_dict(emails_data, orient='columns')
emails_df.index = data.index

Then, we will concatenate the `data` and `emails_df` dataframes to create our new `data` dataframe.

In [62]:
data = pd.concat([data, emails_df], axis=1)

Let's check some records.

In [68]:
data.head()

Date  \
Message-ID                                                          
<18782981.1075855378110.JavaMail.evans@thyme> 2001-05-14 23:39:00   
<15464986.1075855378456.JavaMail.evans@thyme> 2001-05-04 20:51:00   
<24216240.1075855687451.JavaMail.evans@thyme> 2000-10-18 10:00:00   
<13505866.1075863688222.JavaMail.evans@thyme> 2000-10-23 13:13:00   
<30922949.1075863688243.JavaMail.evans@thyme> 2000-08-31 12:07:00   

                                                                    From  \
Message-ID                                                                 
<18782981.1075855378110.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<15464986.1075855378456.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<24216240.1075855687451.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<13505866.1075863688222.JavaMail.evans@thyme>  (phillip.allen@enron.com)   
<30922949.1075863688243.JavaMail.evans@thyme>  (phillip.allen@enron.com)   

                                                                      To  \
Message-ID                                                                 
<18782981.1075855378110.JavaMail.evans@thyme>     (tim.belden@enron.com)   
<15464986.1075855378456.JavaMail.evans@thyme>  (john.lavorato@enron.com)   
<24216240.1075855687451.JavaMail.evans@thyme>   (leah.arsdall@enron.com)   
<13505866.1075863688222.JavaMail.evans@thyme>    (randall.gay@enron.com)   
<30922949.1075863688243.JavaMail.evans@thyme>     (greg.piper@enron.com)   

                                                 Subject           X-From  \
Message-ID                                                                  
<18782981.1075855378110.JavaMail.evans@thyme>             Phillip K Allen   
<15464986.1075855378456.JavaMail.evans@thyme>        Re:  Phillip K Allen   
<24216240.1075855687451.JavaMail.evans@thyme>   Re: test  Phillip K Allen   
<13505866.1075863688222.JavaMail.evans@thyme>             Phillip K Allen   
<30922949.1075863688243.JavaMail.evans@thyme>  Re: Hello  Phillip K Allen   

                                                                                            X-To  \
Message-ID                                                                                         
<18782981.1075855378110.JavaMail.evans@thyme>           Tim Belden <Tim Belden/Enron@EnronXGate>   
<15464986.1075855378456.JavaMail.evans@thyme>  John J Lavorato <John J Lavorato/ENRON@enronXg...   
<24216240.1075855687451.JavaMail.evans@thyme>                                   Leah Van Arsdall   
<13505866.1075863688222.JavaMail.evans@thyme>                                      Randall L Gay   
<30922949.1075863688243.JavaMail.evans@thyme>                                         Greg Piper   

                                              X-cc X-bcc  \
Message-ID                                                 
<18782981.1075855378110.JavaMail.evans@thyme>              
<15464986.1075855378456.JavaMail.evans@thyme>              
<24216240.1075855687451.JavaMail.evans@thyme>              
<13505866.1075863688222.JavaMail.evans@thyme>              
<30922949.1075863688243.JavaMail.evans@thyme>              

                                                                                        X-Folder  \
Message-ID                                                                                         
<18782981.1075855378110.JavaMail.evans@thyme>  \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Se...   
<15464986.1075855378456.JavaMail.evans@thyme>  \Phillip_Allen_Jan2002_1\Allen, Phillip K.\'Se...   
<24216240.1075855687451.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   
<13505866.1075863688222.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   
<30922949.1075863688243.JavaMail.evans@thyme>    \Phillip_Allen_Dec2000\Notes Folders\'sent mail   

                                              X-Origin      ...        \
Message-ID                                                  ...         
<1